File: November_Leetcode_Challenge_2022/Day3/Longest_Palindrome_by_Concatenating_Two_Letter_Words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
     map<string,int>mp;
     int len = 0;
     int ans = 0;
     bool flag = false;
     for(auto word : words)
     {
         mp[word]++;
     }
     for(auto x : mp)
     {
         if(x.first[0] == x.first[1])
         {
             if(x.second % 2 == 1)
             {
                 flag = true;
                 len+=(x.second-1);
             }
             else
             {
                 len+=mp[x.first];
             }
              mp.erase(x.first);
         }
         else
         {
             string str = x.first;
             reverse(str.begin(),str.end());
             if(mp.find(str)!=mp.end())
             {
                 ans+=min(mp[str],x.second)*4;
             }
             mp.erase(str);
             mp.erase(x.first);
         }
     }
        if(flag)
         {
           return ans+len*2+2;  
         }
         return ans+len*2;
    }
};

File: November_Leetcode_Challenge_2022/Day3/Longest_Palindrome_by_Concatenating_Two_Letter_Words.java
class Solution {
    public int longestPalindrome(String[] words) {
        HashMap<String,Integer> map= new HashMap<>();
        for(String word:words){
            map.put(word,map.getOrDefault(word,0)+1);
        }
        int maxcount=0;
        int count=0;
        boolean onetime=true;
         for (Map.Entry<String, Integer> e : map.entrySet()){
             String tempstring=e.getKey();
             int tempint=e.getValue();
             if(tempint==0)continue;
             if(tempstring.charAt(0)==tempstring.charAt(1)){
                 maxcount+=2*(tempint/2);
                 if(tempint%2!=0 && onetime){
                     maxcount++;
                     onetime=false;
                 }
                continue;
             }
             StringBuilder reversestring=new StringBuilder(tempstring).reverse();
             if(map.containsKey(reversestring.toString())){
                 count+=Math.min(tempint,map.get(reversestring.toString()));
                 map.put(reversestring.toString(),0);
             }
             //map.remove(tempstring);
         }
            return 4*count+2*maxcount;
    }
}
